Added comment to generate a PR for code review

golliher opened the pull request.

golliher committed:
Added comment to generate a PR for code review

golliher committed:
Commenting all line to allow a code review of the whole file.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+// }
+
+// // InitializeConfig loads our configuration using Viper package.
+// func InitializeConfig() {
```
LoadConfig

golliher replied:
I think you are telling me that you like the name LoadConfig better than the name InitializeConfig.   Is that right?

FWIW,  I only picked the latter because it seemed to be the convention of the Viper author (as seen in the Hugo source code: https://github.com/spf13/hugo/blob/master/commands/hugo.go#L279

If you were telling me something else, I'm sorry, I missed your point.

zellyn replied:
sgtm - I'm not familiar with Viper, but consisted is best.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+// func InitializeConfig() {
+
+// 	viper.SetConfigType("yaml")
+// 	// Set config file
```
Redundant comment.

golliher replied:
Agreed.  I'll fix.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+// 	// Set config file
+// 	viper.SetConfigName("config")
+
+// 	// Add config path
```
ditto

golliher replied:
Agreed. I'll fix.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+
+// 	// Do some flag handling and any complicated config logic
+// 	if !viper.IsSet("ip") || !viper.IsSet("port") {
+// 		fmt.Println("Configuration error.  Both IP and PORT must be set via either config or environment.")
```
fmt.Fprintln(os.Stderr, …)

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+// 	if err != nil {
+// 		log.Fatal(err)
+// 	}
+// 	webpages := string(bytes)
```
`bytes.NewReader(bytes)` below obviates the need for this. (You'll need to rename the variable of course, once you're importing `bytes`.)

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+
+// 	for {
+// 		var p Program
+// 		record, err := r.Read()
```
Probably simpler to use `ReadAll` and then a range loop over the resulting data.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+// 			log.Fatal(err)
+// 		}
+// 		p.URL = record[0]
+// 		p.Duration, err = time.ParseDuration(record[1])
```
Unhandled err. Consider `go vet` or the go metalinter.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+
+// func runProgram(program Program) {
+
+// 	// Does this leak goroutines over time because they are created more than they are consumed?
```
Yes. Launch it once from the main function, and have it loop. Also, this form of looking for keypresses might not actually work. Consider `termbox-go` or equivalent?

golliher replied:
Thanks for the suggestion for how to fix it!   Later I plan to switch to termbox-go when/if I add more console controls.  For now, I'll move to main and launch in a loop -- it's does read keypresses, but it does trigger when return is pressed and that's good enough for me for now.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+// }
+
+// // Control channel to stop running programs immediately (yes, global)
+// var abort = make(chan struct{})
```
s/abort/skip/ ?

golliher replied:
Naming things is hard.  skip works for me too.  I'll change to it.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+
+// // Control channel to stop running programs immediately (yes, global)
+// var abort = make(chan struct{})
+// var pause bool = false
```
Need mutex to synchronize accesses to this. Consider building with `-race` and seeing if the race detector notices.

golliher replied:
`go build -race` is silent.  I'm not sure I understand the need for a mutex.  I'll have to study this some more.

golliher replied:
Hmm after I moved the keyboard reading into a goroutine launch from main, I do start to get races detected.   This is an areas I'll be studying until I really understand.

zellyn replied:
`go build -race` just builds a binary with race-detection enabled. You won't see output until you actually run it. The nice thing about the Go race detector is that (as far as I am aware) it does not have false positives. It can miss existing races, but if it tells you it sees a race, then actual racing ocurred.

zellyn commented on main.go:
```diff
@@ -1,235 +1,237 @@
+
+// 	go LoadAndRunLoop()
+
+// 	r := mux.NewRouter()
```
For simple uses like this, you could also consider using the Default router.

golliher replied:
That's fair; I think I had originally started down a path of a more elaborate syntax for HTTP control of gotator and then ket the 3rd party router when I retreated to a simpler design.

golliher commented on main.go:
```diff
@@ -1,235 +1,237 @@
+// 	// Set config file
+// 	viper.SetConfigName("config")
+
+// 	// Add config path
```
https://github.com/spf13/hugo/blob/master/commands/hugo.go#L279

golliher committed:
Changes based on code review.

I have not yet adressed all the suggestions from code review; but these
are the easy ones.
